core/src/main/scala/com/avast/clients/rabbitmq/DefaultRabbitMQConnection.scala

```diff
@@ -117,30 +127,28 @@ class DefaultRabbitMQConnection[F[_]: Effect](connection: ServerConnection,
   }
 
   private def taskWithChannel[A](f: ServerChannel => Task[A]): Task[A] = {
-    Task {
-      val ch = createChannel()
-
-      try {
-        f(ch)
-          .doOnFinish { e =>
-            e.foreach(logger.debug(s"Error while executing action with channel $ch", _))
-            Task(ch.close()).executeOn(blockingScheduler)
+    Task
+      .fromEffect(createChannel())
+      .map { channel =>
+        try {
+          f(channel).doOnFinish { e =>
+            e.foreach(logger.debug(s"Error while executing action with channel $channel", _))
+            Task(channel.close()).executeOn(blockingScheduler)
           }
-      } catch {
-        case NonFatal(e) =>
-          logger.debug(s"Error while executing action with channel $ch", e)
+        } catch {
+          case NonFatal(e) =>
+            logger.debug(s"Error while executing action with channel $channel", e)
 
-          Task(ch.close())
-            .executeOn(blockingScheduler)
-            .flatMap(_ => Task.raiseError(e))
+            Task(channel.close()).executeOn(blockingScheduler).flatMap(_ => Task.raiseError(e))
+        }
       }
-    }.flatten // surrounded with Task.apply to catch possible errors when creating the channel
-    .asyncBoundary
+      .flatten // surrounded with Task.apply to catch possible errors when creating the channel
+      .asyncBoundary
   }
 
   /** Closes this factory and all created consumers and producers.
     */
-  override def close(): Unit = {
+  override def close(): F[Unit] = F.delay {
     closeables.foreach(_.close())
```

Comment:
What if exception is thrown here? Shouldn't you be more careful and close the `connection` anyway?